docs/proposals/kubelet-cri-logging.md

```diff
@@ -0,0 +1,262 @@
+collection daemon, `fluentd`. To facilitate the log collection, kubelet creates
+symbolic links to all the docker containers logs under `/var/log/containers`
+with pod and container metadata embedded in the filename.
+
+```
+/var/log/containers/<pod_name>_<pod_namespace>_<container_name>-<container_id>.log`
+```
+
+The fluentd daemon watches the `/var/log/containers/` directory and extract the
+metadata associated with the log from the path. Note that this integration
+requires kubelet to know where the container runtime stores the logs, and will
+not be directly applicable to CRI.
+
+
+## Requirements
+
+
+   1. **Allow log collectors to integrate with kubernetes without knowledge of
+        the underlying container runtime.**
+
+   2. **Allow kubelet to manage the lifecycle of the logs to pave the way for
+        better disk management in the future.** This implies that the lifecycle
+        of containers and their logs need to be decoupled.
+   3. **Provide ways for CRI-compliant runtimes to support all existing logging
+        features, i.e., `kubectl logs`.**
+
+Requirement (1) is crucial to the kubernetes' extensibility and usability across
+all deployments. Requirement (2) stems from the fact that disk is managed by
+kubelet as a node-level resource (not per-pod) today, hence it is difficult to
+delegate to the runtime by enforcing per-pod disk quota policy. In addition,
+container disk quota is not well supported yet, and such limitation may not
+even be well-perceived by users. As for requirement (3), even though `kubectl
+logs --since` is only supported today for a limited set of log drivers, this is
+an important usability tool for a fresh, basic kubernetes cluster. We should
+ensure that runtimes have the opportunity to support such features.
+
+
+## Proposed solution
+
+This proposal intends to satisfy the requirements by
+
+  1. Enforce where the container logs should be stored on the host
+     filesystem. Both kubelet and the log collector can interact with
+     the log files directly.
+
+  2. Ask the runtime to decorate the logs in a format that kubelet understands.
+
+**Log directories and structures**
+
+Kubelet will be configured with a root directory (e.g., `/var/log/pods` or
+`/var/lib/kubelet/logs/) to store all container logs. Below is an example of a
+path to the log of a container in a pod.
+
+```
+/var/log/pods/<podUID>/<containerName>_<instance#>.log
+```
+
+In CRI, this is implemented by setting the pod-level log directory when
+creating the pod sandbox, and passing the relative container log path
+when creating a container.
+
+```
+PodSandboxConfig.LogDirectory: /var/log/pods/<podUID>/
+ContainerConfig.LogPath: <containerName>_<instance#>.log
+```
+
+Because kubelet determines where the logs are stores and can access them
+directly, this meets requirement (1). As for requirement (2), the log collector
+can easily extracts all necessary metadata (e.g., pod UID, container name) from
```

Comment:
..all necessary metadata...   I would say this is misleading given for the openshift aggregated logging solution we consider the pod labels 'necessary' but are unable to 'access' them without a call to the api server.